x/evm/keeper/grpc_query.go

```diff
@@ -73,66 +76,69 @@ func (q Querier) BurnerInfo(c context.Context, req *types.BurnerInfoRequest) (*t
 	return nil, status.Error(codes.NotFound, "unknown address")
 }
 
-func getExecuteDataAndSigs(ctx sdk.Context, s types.Signer, commandBatch types.CommandBatch) ([]byte, []types.Signature, error) {
-	id := hex.EncodeToString(commandBatch.GetID())
-	sig, sigStatus := s.GetSig(ctx, id)
-	if sigStatus != tss.SigStatus_Signed {
-		return nil, nil, fmt.Errorf("could not find a corresponding signature for sig ID %s", id)
+func getProof(ctx sdk.Context, commandBatch types.CommandBatch, s types.MultisigKeeper) (types.AuthData, error) {
+	signature := commandBatch.GetSignature().(*multisigtypes.MultiSig)
+
+	key, ok := s.GetKey(ctx, signature.KeyID)
+	if !ok {
+		return types.AuthData{}, fmt.Errorf("key %s not found", signature.KeyID)
 	}
 
-	switch sig := sig.GetSig().(type) {
-	case *tss.Signature_SingleSig_:
-		commandBatchSig, err := getCommandBatchSig(sig.SingleSig.SigKeyPair, commandBatch.GetSigHash())
-		if err != nil {
-			return nil, nil, err
+	operators := slices.Map(key.GetParticipants(), func(val sdk.ValAddress) types.Operator {
+		pubKey := funcs.MustOk(key.GetPubKey(val))
+		pk := funcs.Must(pubKey.GetECDSAPubKey())
+		weight := key.GetWeight(val)
+
+		addr := crypto.PubkeyToAddress(pk)
+		sig, _ := signature.Sigs[val.String()]
+		if sig != nil {
+			sig = funcs.Must(types.NewSignature(sig)).ToHomesteadSig()
 		}
 
-		executeData, err := types.CreateExecuteDataSinglesig(commandBatch.GetData(), commandBatchSig)
-		if err != nil {
-			return nil, nil, fmt.Errorf("could not create transaction data: %s", err)
+		return types.Operator{
+			Address:   types.Address(addr),
+			Signature: sig,
+			Weight:    weight,
 		}
+	})
 
-		return executeData, []types.Signature{commandBatchSig}, nil
-	case *tss.Signature_MultiSig_:
-		sigKeyPairs := types.SigKeyPairs(sig.MultiSig.SigKeyPairs)
-		sort.Stable(sigKeyPairs)
+	//optimize sigs
+	sort.SliceStable(operators, func(i, j int) bool {
+		return operators[i].GetWeight().GT(operators[j].GetWeight())
+	})
 
-		key, ok := s.GetKey(ctx, commandBatch.GetKeyID())
-		if !ok {
-			return nil, nil, fmt.Errorf("key %s not found", commandBatch.GetKeyID())
+	cumWeight := sdk.ZeroUint()
+	for _, o := range operators {
+		if cumWeight.GTE(key.GetMinPassingWeight()) {
+			o.Signature = nil
 		}
+		cumWeight = cumWeight.Add(o.Weight)
```

Comment:
We should only add weights when signature is present
```suggestion
		if cumWeight.GTE(key.GetMinPassingWeight()) {
			o.Signature = nil
		} else if o.Signature != nil {
		        cumWeight = cumWeight.Add(o.Weight)
		}
```